### PR TITLE
[BEAM-1530] Cleanup BigQueryIO, and allow for value-based tables

### PR DESCRIPTION
**Pull request by reuvenlax**
BigQueryIO.Write takes in a user type, which the user can convert to a TableRow.

Table functions now take a ValueInSingleWindow, instead of just the window.

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 1]
```diff
@@ -120,6 +120,7 @@
 import org.apache.beam.sdk.util.Reshuffle;
 import org.apache.beam.sdk.util.SystemDoFnInternal;
 import org.apache.beam.sdk.util.Transport;
+import org.apache.beam.sdk.util.WindowedValue;
```
Comment (tgroh) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: We don't want to expose `WindowedValues` to users directly - it's the compressed representation that doesn't really exist for Pipeline authors.

We've added a `ValueInSingleWindow` in the testing package, which could be pulled up to the `value` package and used here instead.

**Comment by jkff** [comment]
Self-mention @jkff  to put this in my mentioned PRs

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java [hunk 2]
```diff
@@ -1119,42 +1121,45 @@ public void testStreamingWriteWithWindowFn() throws Exception {
...
-            return value.toString();
+          public String apply(Integer i) {
+            return Integer.toString(i % 5);
           }
         }
     );
 
-    SerializableFunction<BoundedWindow, String> tableFunction =
-        new SerializableFunction<BoundedWindow, String>() {
+    SerializableFunction<ValueInSingleWindow<Integer>, String> tableFunction =
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: Actually this business with ValueInSingleWindow looks a little awkward. I can't suggest something much better (especially given that we want to later extend it to support side inputs and such), so I am personally OK with it, but please get an additional person to sign off on it, e.g. Davor.

File: examples/java/src/main/java/org/apache/beam/examples/complete/StreamingWordExtract.java [hunk 3]
```diff
@@ -136,7 +136,8 @@ public static void main(String[] args) throws IOException {
         .apply(ParDo.of(new ExtractWords()))
         .apply(ParDo.of(new Uppercase()))
         .apply(ParDo.of(new StringToRowConverter()))
-        .apply(BigQueryIO.write().to(tableSpec)
+        .apply(BigQueryIO.<TableRow>write().to(tableSpec)
```
Comment (reuvenlax) [inline examples/java/src/main/java/org/apache/beam/examples/complete/StreamingWordExtract.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 4]
```diff
@@ -434,6 +436,19 @@ public TableReference apply(String jobUuid) {
     return NestedValueProvider.of(table, new TableRefToTableSpec());
   }
 
+
+  /**
+   * A formatting function that maps a TableRow to itself. This allows sending a
+   * {@code PCollection<TableRow>} directly to BigQueryIO.Write.
+   */
+  public static final SerializableFunction<TableRow, TableRow> IDENTITY_FORMATTER =
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 5]
```diff
@@ -1371,8 +1401,8 @@ private static String getExtractDestinationUri(String extractDestinationDir) {
    *   }
    * }}</pre>
    */
-  public static Write write() {
-    return new AutoValue_BigQueryIO_Write.Builder()
+  public static <T> Write<T> write() {
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 6]
```diff
@@ -1500,21 +1533,21 @@ private void ensureToNotCalledYet() {
     }
 
     /**
-     * Creates a write transformation for the given table specification.
-     *
-     * <p>Refer to {@link #parseTableSpec(String)} for the specification format.
+     * @deprecated Should be private. Instead, use one of the factory methods in
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Fixed

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 7]
```diff
@@ -1525,43 +1558,47 @@ public Write to(ValueProvider<String> tableSpec) {
     }
 
     /**
-     * Creates a write transformation from a function that maps windows to table specifications.
-     * Each time a new window is encountered, this function will be called and the resulting table
-     * will be created. Records within that window will be written to the associated table.
+     * Returns a copy of this write transformation, but writing to the specified table. Refer to
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 8]
```diff
@@ -1572,16 +1609,22 @@ public TableReference apply(BoundedWindow value) {
...
     }
 
     /**
-     * Like {@link #withSchema(TableSchema)}, but with a {@link ValueProvider}.
+     * Returns a copy of this write transformation, but using the specified function to determine
+     * which table to write to for each window.
+     *
+     * <p>Does not modify this object.
+     *
+     * <p>{@code tableRefFunction} should be deterministic. When given the same window value, it
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Fixed

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 9]
```diff
@@ -1661,6 +1704,9 @@ public void validate(PCollection<TableRow> input) {
           "Cannot set both a table reference and a table function for a BigQueryIO.Write"
               + " transform");
 
+      checkArgument(getFormatFunction() != null,
+                    "A function must be provided to convert type into a TableRow");
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java (hunk 2 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: @tgroh suggested it as well. However, I'll double check with Davor as well.

File: examples/java8/src/main/java/org/apache/beam/examples/complete/game/utils/WriteToBigQuery.java [hunk 10]
```diff
@@ -121,10 +121,9 @@ protected TableSchema getSchema() {
   public PDone expand(PCollection<InputT> teamAndScore) {
     return teamAndScore
       .apply("ConvertToRow", ParDo.of(new BuildRowFn()))
-      .apply(BigQueryIO.write()
-                .to(getTable(teamAndScore.getPipeline(),
-                    tableName))
-                .withSchema(getSchema())
+      .apply(BigQueryIO.writeTableRows()
+                .to(getTable(teamAndScore.getPipeline(), tableName))
+          .withSchema(getSchema())
```
Comment (jkff) [inline examples/java8/src/main/java/org/apache/beam/examples/complete/game/utils/WriteToBigQuery.java]: Reformat this thing

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 11]
```diff
@@ -195,12 +197,17 @@
...
  *
- * quotes.apply(BigQueryIO.Write
+ * quotes.apply(BigQueryIO.writeTableRows()
  *     .to("my-project:output.output_table")
  *     .withSchema(schema)
  *     .withWriteDisposition(BigQueryIO.Write.WriteDisposition.WRITE_TRUNCATE));
  * }</pre>
  *
+ * <p>BigQueryIO.Write can also take in a user-defined type, as long as a function is provided to
+ * turn
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Reformat this part

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 12]
```diff
@@ -2769,11 +2823,12 @@ public PDone expand(PCollection<TableRow> input) {
       // BigQuery built-in best effort de-dup mechanism.
 
       // To use this mechanism, each input TableRow is tagged with a generated
-      // unique id, which is then passed to BigQuery and used to ignore duplicates.
+      // unique id, which is then passed to BigQuery and usedTableRowInfo to ignore duplicates.
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Accidental paste?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 13]
```diff
@@ -2769,11 +2823,12 @@ public PDone expand(PCollection<TableRow> input) {
...
 
       PCollection<KV<ShardedKey<String>, TableRowInfo>> tagged = input.apply(ParDo.of(
-          new TagWithUniqueIdsAndTable(input.getPipeline().getOptions().as(BigQueryOptions.class),
-              write.getTable(), write.getTableRefFunction())));
+          new TagWithUniqueIdsAndTable<T>(input.getPipeline().getOptions().as(
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Reformat this statement - the line breaking of the first argument looks very strange

File: examples/java8/src/main/java/org/apache/beam/examples/complete/game/utils/WriteToBigQuery.java (hunk 10 above)
Comment (reuvenlax) [inline examples/java8/src/main/java/org/apache/beam/examples/complete/game/utils/WriteToBigQuery.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 11 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 12 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Fixed

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 13 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done

**Comment by reuvenlax** [comment]
@tgroh does this look good to you? If so, could you help with the merge?

File: examples/java8/src/main/java/org/apache/beam/examples/complete/game/utils/WriteToBigQuery.java [hunk 14]
```diff
@@ -121,10 +121,8 @@ protected TableSchema getSchema() {
   public PDone expand(PCollection<InputT> teamAndScore) {
     return teamAndScore
       .apply("ConvertToRow", ParDo.of(new BuildRowFn()))
-      .apply(BigQueryIO.write()
-                .to(getTable(teamAndScore.getPipeline(),
-                    tableName))
-                .withSchema(getSchema())
+      .apply(BigQueryIO.writeTableRows().to(getTable(teamAndScore.getPipeline(), tableName))
```
Comment (tgroh) [inline examples/java8/src/main/java/org/apache/beam/examples/complete/game/utils/WriteToBigQuery.java]: This formatting is all wacky

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: @davorbonaci

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 15]
```diff
@@ -195,12 +197,16 @@
  * fields.add(new TableFieldSchema().setName("quote").setType("STRING"));
  * TableSchema schema = new TableSchema().setFields(fields);
  *
- * quotes.apply(BigQueryIO.Write
+ * quotes.apply(BigQueryIO.writeTableRows()
```
Comment (tgroh) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Line 191 is no longer correct

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 16]
```diff
@@ -1352,8 +1371,9 @@ private static String getExtractDestinationUri(String extractDestinationDir) {
   /////////////////////////////////////////////////////////////////////////////
 
   /**
-   * A {@link PTransform} that writes a {@link PCollection} containing {@link TableRow TableRows}
-   * to a BigQuery table.
+   * A {@link PTransform} that writes a {@link PCollection} containing objects of type {@code T}
```
Comment (tgroh) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: I would reword this approximately as follows:

"A `PTransform` that writes a `PCollection` to a BigQuery table. A Formatting function must be provided to convert each input element into a `TableRow` using `Write#withFormatFunction(SerializableFunction)`"

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 17]
```diff
@@ -1756,10 +1798,13 @@ public String apply(String input) {
 
       PCollection<TableRow> inputInGlobalWindow =
           input.apply(
-              Window.<TableRow>into(new GlobalWindows())
+              Window.<T>into(new GlobalWindows())
                   .triggering(DefaultTrigger.of())
-                  .discardingFiredPanes());
+                  .discardingFiredPanes())
+              .apply(MapElements.via(getFormatFunction())
```
Comment (tgroh) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Can we skip over this `MapElements` if the result of `getFormatFunction()` equals `IDENTITY_FORMATTER`?

It's not unreasonable to not, as we get the appropriate behavior out, but we get to not include the extra step.

File: examples/java8/src/main/java/org/apache/beam/examples/complete/game/utils/WriteToBigQuery.java (hunk 14 above)
Comment (reuvenlax) [inline examples/java8/src/main/java/org/apache/beam/examples/complete/game/utils/WriteToBigQuery.java]: Fixed

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 15 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Fixed

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 16 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 17 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done - it does require some casting.

**Comment by tgroh** [comment]
LGTM. Would you like to rebase, or should I?

**Comment by reuvenlax** [comment]
Rebased against origin/master.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 18]
```diff
@@ -527,15 +546,15 @@ public Read fromQuery(String query) {
...
+     * Same as {@code fromQuery(String)}, but with a {@link ValueProvider}.
      */
     public Read fromQuery(ValueProvider<String> query) {
       ensureFromNotCalledYet();
       return toBuilder().setQuery(query).setFlattenResults(true).setUseLegacySql(true).build();
     }
 
     /**
-     * Reads a BigQuery table specified as a {@link TableReference} object.
+     * Like {@link #fromQuery(String)}, but from a {@link ValueProvider}.
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: This is incorrect

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 19]
```diff
@@ -1531,43 +1562,47 @@ public Write to(ValueProvider<String> tableSpec) {
     }
 
     /**
-     * Creates a write transformation from a function that maps windows to table specifications.
-     * Each time a new window is encountered, this function will be called and the resulting table
-     * will be created. Records within that window will be written to the associated table.
+     * Returns a copy of this write transformation, but with the specified table function. The
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: While you're at it - clean up these comments? "Returns a copy..." and "Does not modify.." are unnecessary.

**Comment by jkff** [comment]
Can you squash things into more self-contained commits (I suppose 2 commits: 1) move ValueInSingleWindow 2) the BigQueryIO changes per se)?

**Comment by reuvenlax** [comment]
@jkff commits are squashed
